core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java

```diff
@@ -897,7 +906,12 @@ public MergeWriterNode plan(Merge merge)
             Symbol symbol = planWithPresentColumn.getFieldMappings().get(fieldIndex);
             projectionAssignmentsBuilder.putIdentity(symbol);
         }
-        projectionAssignmentsBuilder.putIdentity(uniqueIdSymbol);
+
+        // define a unique id come from target or source, if join match will use target, otherwise use source
```

Comment:
Please explain in the comment why using the unique ID from source side does not break semantics, and that it is done to improve performance of MarkDistinct.

Reply:
Also, after this change, the `isNull` check is redundant:
```
Logical.and(
      not(metadata, isDistinctSymbol.toSymbolReference()),
      not(metadata, new IsNull(uniqueIdSymbol.toSymbolReference()))),
```